### PR TITLE
Account Transactions: Encrypted transfer transactions

### PR DESCRIPTION
**Pull request by tds-1**
This PR includes following account transactions:
1. Transfer to public (https://github.com/Concordium/concordium-java-sdk/pull/140)
2. Transfer to encrypted. (https://github.com/Concordium/concordium-java-sdk/pull/138)
3. Encrypted amount transfer.
4. Encrypted amount transfer with memo

### CODE REVIEW

File: concordium-sdk/src/main/java/com/concordium/sdk/crypto/encryptedtransfers/AggregatedDecryptedAmount.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+import com.concordium.sdk.responses.accountinfo.AccountEncryptedAmount;
+import lombok.Builder;
+import lombok.Data;
+import lombok.extern.jackson.Jacksonized;
+
+@Jacksonized
+@Builder
+@Data
+public class AggregatedDecryptedAmount {
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/crypto/encryptedtransfers/AggregatedDecryptedAmount.java:11]: Documentation

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/accountinfo/AccountInfo.java [hunk 2]
```diff
@@ -7,107 +7,94 @@
...
 import java.util.Map;
 import java.util.Objects;
 
 /**
  * AccountInfo
  */
-@Getter
-@ToString
 @EqualsAndHashCode(onlyExplicitlyIncluded = true)
+@Data
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/responses/accountinfo/AccountInfo.java:24]: Is this related to the actual PR and the new transaction types? 

It's ok for now. But please refrain from doing too much other than what is required from the PR as it makes it harder to review the PR. Especially for larger ones like this.

File: concordium-sdk/src/main/java/com/concordium/sdk/responses/accountinfo/AccountEncryptedAmount.java [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+import lombok.Builder;
+import lombok.Data;
+import lombok.extern.jackson.Jacksonized;
+
+import java.util.List;
+
+@Data
+@Jacksonized
+@Builder
+public final class AccountEncryptedAmount {
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/responses/accountinfo/AccountEncryptedAmount.java:14]: Documentation

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedAmount.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    private EncryptedAmount(final byte[] bytes) {
+        this.bytes = bytes;
+    }
+
+    @JsonCreator
+    public static EncryptedAmount from(String hexKey) {
+        try {
+            return new EncryptedAmount(Hex.decodeHex(hexKey));
+        } catch (DecoderException e) {
+            throw new IllegalArgumentException("Cannot create AccountEncryptedAmount", e);
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedAmount.java:30]: Why `AccountEncryptedAmount` and not just `EncryptedAmount` ?

**Review by EmilLa1 — CHANGES_REQUESTED** [review]
I haven't looked in depth of everything, however I would appreciate if you could look at the comments I've made and applied them where applicable. 
I.e. documentation and _wrong_ error messages (json deserialization).

Please remove the `crypto-jni/src/tables_bytes.bin` file, it shouldn't be used anymore no?

File: concordium-sdk/src/main/java/com/concordium/sdk/crypto/CryptoJniNative.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package com.concordium.sdk.crypto;
+
+public class CryptoJniNative {
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/crypto/CryptoJniNative.java:3]: Documentation on the functions here. In particular the return values.

File: concordium-sdk/src/main/java/com/concordium/sdk/crypto/CryptoJniResultCode.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+import lombok.Getter;
+
+public enum CryptoJniResultCode {
+    SUCCESS(0),
+    JSON_DESERIALIZATION_ERROR(1, "Json Deserialization error."),
+    UTF8_ERROR(2, "Incorrect Input String."),
+    NATIVE_CONVERSION_ERROR(3, "Native Conversion Error."),
+    AMOUNT_DECRYPTION_ERROR(4, "Input Encrypted Amount could not be Decrypted."),
+    PAYLOAD_CREATION_ERROR(5, "JNI Output Payload could not be created"),
+    ERROR_UNKNOWN_RESULT_CODE(100, "Unknown EncryptedTransfersResultCode");
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply concordium-sdk/src/main/java/com/concordium/sdk/crypto/CryptoJniResultCode.java]: Is it a generic unknown result code? Or is it solely encrypted transfers related? If the former I think it makes sense to have it kind of special e.g. `-1`. But if it is solely used as part of the encrypted transfers then I don't see why it's special. 
The name of the enum should reflect this.

File: concordium-sdk/src/main/java/com/concordium/sdk/crypto/encryptedtransfers/GlobalContext.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+import com.concordium.sdk.crypto.bulletproof.BulletproofGenerators;
+import com.concordium.sdk.crypto.pedersencommitment.PedersenCommitmentKey;
+import com.concordium.sdk.responses.cryptographicparameters.CryptographicParameters;
+import lombok.Builder;
+import lombok.Data;
+
+@Data
+@Builder
+class GlobalContext {
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/crypto/encryptedtransfers/GlobalContext.java:11]: Documentation. When is this used, what is the lifetime? e.g..

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferTransaction.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+public class EncryptedTransferTransaction extends AbstractTransaction {
+
+    /**
+     *  Data that will go onto an encrypted amount transfer.
+     */
+    private final EncryptedAmountTransferData data;
+    /**
+     * The account address to which the transfer will be sent.
+     */
+    private final AccountAddress to;
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferTransaction.java:21]: Please name it `receiver` so it is aligned with e.g. `Transfer`.

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferTransaction.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+                AccountNonce nonce,
+                Expiry expiry,
+                TransactionSigner signer,
+                AccountAddress to,
+                EncryptedAmountTransferData data) throws TransactionCreationException {
+
+            Transaction.verifyAccountTransactionHeaders(sender, nonce, expiry, signer);
+
+            if (Objects.isNull(to)) {
+                throw TransactionCreationException.from(new IllegalArgumentException("To address cannot be null"));
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferTransaction.java:114]:
```suggestion
                throw TransactionCreationException.from(new IllegalArgumentException("Receiver address cannot be null"));
```

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferTransaction.java (hunk 8 above)
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferTransaction.java:21]:
```suggestion
    private final AccountAddress receiver;
```

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferWithMemo.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+public final class EncryptedTransferWithMemo extends Payload {
+    /**
+     *  Data that will go onto an encrypted amount transfer.
+     */
+    private final EncryptedAmountTransferData data;
+
+    /**
+     * Account Address of the sender.
+     */
+    private final AccountAddress to;
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferWithMemo.java:25]: Same here

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferWithMemoTransaction.java [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+    /**
+     *  Data that will go onto an encrypted amount transfer.
+     */
+    private final EncryptedAmountTransferData data;
+
+    /**
+     * Account Address of the sender.
+     */
+    private final AccountAddress to;
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferWithMemoTransaction.java:22]: Same here

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferWithMemoTransaction.java [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    private final Memo memo;
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferWithMemoTransaction.java:23]: Documentation

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferWithMemoTransaction.java [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+                Expiry expiry,
+                TransactionSigner signer,
+                AccountAddress to,
+                EncryptedAmountTransferData data,
+                Memo memo) throws TransactionCreationException {
+
+            Transaction.verifyAccountTransactionHeaders(sender, nonce, expiry, signer);
+
+            if (Objects.isNull(to)) {
+                throw TransactionCreationException.from(new IllegalArgumentException("To address cannot be null"));
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferWithMemoTransaction.java:120]: Same here

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferWithMemoTransaction.java [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+            }
+            if (Objects.isNull(memo)) {
+                throw TransactionCreationException.from(new IllegalArgumentException("To address cannot be null"));
```
Comment (EmilLa1) [inline concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferWithMemoTransaction.java:123]: This error message is not correct is it?

**Review by EmilLa1 — CHANGES_REQUESTED** [review]
Looks mostly good. 
A few documentation / naming patterns that I think we need to reconsider.

File: concordium-sdk/src/main/java/com/concordium/sdk/crypto/encryptedtransfers/GlobalContext.java (hunk 7 above)
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/crypto/encryptedtransfers/GlobalContext.java:15]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/crypto/CryptoJniNative.java (hunk 5 above)
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/crypto/CryptoJniNative.java:3]: done

File: concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferWithMemoTransaction.java (hunk 12 above)
Comment (parv0888) [reply concordium-sdk/src/main/java/com/concordium/sdk/transactions/EncryptedTransferWithMemoTransaction.java:27]: done

**Review by EmilLa1 — APPROVED** [review]
GJ, looks good.

**Pull request merged**